File: C++/78. Subsets.cpp
class Solution {
public:    
    vector<vector<int>> ans;    
    void help(vector<int> v,int c){        
        int n = v.size();        
        vector<int> t;        
        for(int i=0;i<n;i++){
            int x = (c&1);            
            if(x) t.push_back(v[i]);            
            c>>=1;
        }        
        ans.push_back(t);
    }    
    vector<vector<int>> subsets(vector<int>& v) {        
        int n = v.size();        
        for(int i=0;i<(1<<n);i++) help(v,i);        
        return ans;        
    }
};

File: JAVA/78. Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(ans,temp,0,nums);
        return ans;
    }
    public void helper(List<List<Integer>> ans,List<Integer> temp,int start, int[] nums){
        if(start >= nums.length){
            ans.add(new ArrayList<>(temp));
        }
        else{
            temp.add(nums[start]);
            helper(ans,temp,start+1,nums);
            temp.remove(temp.size()-1);
            helper(ans,temp,start+1,nums);
        }
    }
}
